lib/lookup/src/lookup_buf/mod.rs

```diff
@@ -0,0 +1,399 @@
+pub struct LookupBuf {
+    pub segments: VecDeque<SegmentBuf>,
+}
+
+impl<'a> TryFrom<VecDeque<SegmentBuf>> for LookupBuf {
+    type Error = LookupError;
+
+    fn try_from(segments: VecDeque<SegmentBuf>) -> Result<Self, Self::Error> {
+        let retval = LookupBuf { segments };
+        retval.is_valid()?;
+        Ok(retval)
+    }
+}
+
+impl Display for LookupBuf {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let mut peeker = self.segments.iter().peekable();
+        let mut next = peeker.next();
+        let mut maybe_next = peeker.peek();
+        while let Some(segment) = next {
+            match segment {
+                SegmentBuf::Field(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"{}."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "{}", segment)?,
+                },
+                SegmentBuf::Index(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"[{}]."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "[{}]", segment)?,
+                },
+                SegmentBuf::Coalesce(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"{}."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "{}", segment)?,
+                },
+            }
+            next = peeker.next();
+            maybe_next = peeker.peek();
+        }
+        Ok(())
+    }
+}
+
+impl LookupBuf {
+    /// Creates an lookup to the root
+    pub fn root() -> Self {
+        Self {
+            segments: VecDeque::new(),
+        }
+    }
+
+    /// Get from the internal list of segments.
+    #[instrument(level = "trace")]
+    pub fn get(&mut self, index: usize) -> Option<&SegmentBuf> {
+        self.segments.get(index)
+    }
+
+    /// Push onto the internal list of segments.
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_back(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_back(segment.into());
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_back(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_back()
+    }
+
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_front(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_front(segment.into())
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_front(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_front()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn iter(&self) -> std::collections::vec_deque::Iter<'_, SegmentBuf> {
+        self.segments.iter()
+    }
+
+    #[instrument(level = "trace")]
```

Comment:
I remember you starting a discussion about the excessive use of instrumentation, and wanting to trim that down. I'd say this (and others) qualifies for such an excessive use, since you're tracing Rust's `len` performance here, and nothing else.